apps/arweave/src/ar_peers.erl

```diff
@@ -552,59 +641,134 @@ rank_peers(ScoredPeers) ->
 			[],
 			GroupedBySubnet
 		),
-	[Peer || {Peer, _} <- lists:sort(fun({_, S1}, {_, S2}) -> S1 >= S2 end,
-			ScoredSubnetPeers)].
+	[Peer || {Peer, _} <- lists:sort(
+		fun({_, S1}, {_, S2}) -> S1 >= S2 end,
+		ScoredSubnetPeers
+	)].
 
-check_external_peer(Peer) ->
-	IsLoopbackIP = is_loopback_ip(Peer),
+set_ranked_peers(Rating, Peers) ->
+	ets:insert(?MODULE, {{peers, Rating}, lists:sublist(Peers, ?MAX_PEERS)}).
+
+check_peer(Peer) ->
+	check_peer(Peer, not is_loopback_ip(Peer)).
+check_peer(Peer, IsPeerScopeValid) ->
 	IsBlacklisted = lists:member(Peer, ?PEER_PERMANENT_BLACKLIST),
 	IsBanned = ar_blacklist_middleware:is_peer_banned(Peer) == banned,
-	case {IsLoopbackIP, IsBlacklisted, IsBanned} of
-		{true, _, _} ->
-			reject;
-		{_, true, _} ->
-			reject;
-		{_, _, true} ->
-			reject;
-		_ ->
-			ok
+	case IsPeerScopeValid andalso not IsBlacklisted andalso not IsBanned of
+		true ->
+			ok;
+		false ->
+			reject
 	end.
 
-update_rating(Peer, TimeDelta, Size) ->
+update_rating(Peer, IsSuccess) ->
+	update_rating(Peer, undefined, undefined, 1, IsSuccess).
+update_rating(Peer, LatencyMilliseconds, DataSize, Concurrency, false)
+  		when LatencyMilliseconds =/= undefined; DataSize =/= undefined ->
+	%% Don't credit peers for failed requests.
+	update_rating(Peer, undefined, undefined, Concurrency, false);	
+update_rating(Peer, LatencyMilliseconds, DataSize, Concurrency, IsSuccess) ->
 	Performance = get_or_init_performance(Peer),
-	Total = get_total_rating(),
-	#performance{ bytes = Bytes, time = Time,
-			rating = Rating, transfers = N } = Performance,
-	Bytes2 = Bytes + Size,
-	Time2 = Time + TimeDelta / 1000,
-	Performance2 = Performance#performance{ bytes = Bytes2, time = Time2,
-			rating = Rating2 = Bytes2 / (Time2 + 1), failures = 0, transfers = N + 1 },
-	Total2 = Total - Rating + Rating2,
-	may_be_rotate_peer_ports(Peer),
-	ets:insert(?MODULE, [{{peer, Peer}, Performance2}, {rating_total, Total2}]).
-
-get_or_init_performance(Peer) ->
+	
+	#performance{
+		average_bytes = AverageBytes,
+		total_bytes = TotalBytes,
+		average_latency = AverageLatency,
+		total_latency = TotalLatency,
+		average_success = AverageSuccess,
+		lifetime_rating = LifetimeRating,
+		current_rating = CurrentRating,
+		total_transfers = TotalTransfers
+	} = Performance,
+	TotalBytes2 = case DataSize of
+		undefined -> TotalBytes;
+		_ -> TotalBytes + DataSize
+	end,
+	%% AverageBytes is the average number of bytes transferred during the AverageLatency time
+	%% period. In order to approximate the impact of multiple concurrent requests we multiply
+	%% DataSize by the Concurrency value. We do this *only* when updating the AverageBytes
+	%% value so that it doesn't distort the TotalBytes.
+	AverageBytes2 = case DataSize of
+		undefined -> AverageBytes;
+		_ -> calculate_ema(AverageBytes, (DataSize * Concurrency), ?THROUGHPUT_ALPHA)
```

Comment:
What happens if we do not use `Concurrency` here? Do I understand it correctly that when we download from 2 peers at the same time, the contribution to both ratings is doubled?

Reply:
The concurrency value allows us to estimate the true throughput from that peer better. Since we are only considering one data point at a time if we ignore concurrency we underestimate the throughput. 

For example if we make two concurrent requests to a peer:
A. 1kb with a latency if 1second
B. 2kb with a latency if 1second

The true throughput is 3kb per second. 

However if ignore concurrency we'll average each data point and end up with 1.5kb per second. 

When using a concurrency value of 2 we compute the normal data average (1.5kb) and then multiply by 2 to approximate the fact that there are 2 streams of data concurrently and. And we end up with the "correct" 3kb per second

It is just an approximation - it won't always yield exactly the true throughput. But it is much closer than just considering each data point in isolation. This is primarily important for data syncing where we regularly have many concurrent requests going